projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+package main
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/csv"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// define a type of struct
+// (!) ` ` are struct tags, it provides metadata which can be used by Go packages like encoding/json, otherwise the unmarshal will use the field names
+type Player struct {
+	Name      string `json:"name"`
+	HighScore int32  `json:"high_score"`
+}
+
+// [1] JSON
+func parseJSON(file []byte, dataSlice *[]Player) error {
+	// try to "unmarshal" the data from the json file into the jsonDataSlice variable
+	err := json.Unmarshal(file, &dataSlice)
+
+	// if we fail to "unmarshal" the json file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in JSON format")
```

Comment:
What makes an empty list an invalid input to this function?

(Same question for the others :))

Reply:
should i return an empty slice instead? since it was able to unmarshal but was empty(?)